### PR TITLE
updated unittests, stated shortcut functions, resolved #8 and resolved #4

### LINKED ISSUE #4: Implement a cache system for complex queries
Implement some cache functions for common operations (i.e. species content, duplications, etc). For instance, if the pattern contain statements checking node size, you can precompute a dictionary like node2size (take a look at the current Tree.get_cached_content) that is later used. If many things are cached, a single dict could be shared, saving memory. Node2cache = {node: [size, species, leaves] }. Alternatively, we can just keep a node2descendants dictionary, so we calculate any value on the fly but without the need of traversing the tree back and forth many times.

### LINKED ISSUE #8: docstring documentation is missing in TreeMatcher class and functions.

### PR DESCRIPTION
**Pull request by AlishaMechtley**

### CODE REVIEW

**Comment by unode** [comment]
Hi @AlishaMechtley . This pull request has merge conflicts. Can you solve them?

I think it's just because you haven't updated your fork to include the last changes from `etetoolkit/treematcher`.

File: ete_treematcher.py [hunk 1]
```diff
@@ -7,6 +7,10 @@
 
 
 def populate_args(extract_args_p):
+    """
+
+    :param extract_args_p:
```
Comment (jhcepas) [inline ete_treematcher.py:12]: no need for docstrings in internal functions that are not going to be exposed to the user.

File: test_treematcher.py [hunk 2]
```diff
@@ -29,33 +30,39 @@ def test_ete_params(self):
         pattern2 = TreePattern(pattern2, format=8, quoted_node_names=True)
 
         tree = Tree("(hello,(1,2,3)kk)pasa:1;", format=1)
-        match0 = pattern0.find_match(tree, None)
-        match1 = pattern1.find_match(tree, None)[1]
-        match2 = pattern2.find_match(tree, None)[1]
-        self.assertEqual(match0, (False, None))
-        self.assertEqual(match1.name, 'pasa')
-        self.assertEqual(match2.name, 'pasa')
+        match0 = pattern0.find_match(tree, None, maxhits=1)
```
Comment (jhcepas) [inline test_treematcher.py:33]: rather than enforcing maxhits, evaluate the the number of hits is 1

File: treematcher.py [hunk 3]
```diff
@@ -92,39 +130,66 @@ def __str__(self):
...
         if self.cache_flag == 1:
             self.preprocess(tree)
 
         num_hits = 0
+        num = 0
         for node in tree.traverse("preorder"):
+            num += 1
 
             if self.is_match(node, local_vars=local_vars):
+                print num, '- YES'
```
Comment (jhcepas) [inline treematcher.py]: all these print statements are not python 3 compatible

**Pull request merged**